java-checks/src/main/java/org/sonar/java/checks/helpers/Javadoc.java

```diff
@@ -0,0 +1,266 @@
+      this.name = name;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      } else if (o instanceof BlockTagKey) {
+        BlockTagKey other = ((BlockTagKey) o);
+        return tag == other.tag && Objects.equals(name, other.name);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(tag, name);
+    }
+
+    static BlockTagKey of(BlockTag tag, @Nullable String name) {
+      return new BlockTagKey(tag, name);
+    }
+  }
+
+  private static final Tree.Kind[] CLASS_KINDS = PublicApiChecker.classKinds();
+  private static final Tree.Kind[] METHOD_KINDS = PublicApiChecker.methodKinds();
+  private static final List<String> GENERIC_EXCEPTIONS = Arrays.asList("Exception", "java.lang.Exception");
+  private static final Pattern BLOCK_TAG_LINE_PATTERN = Pattern.compile("^@\\S+.*");
+  private static final Set<String> PLACEHOLDERS = ImmutableSet.of("TODO", "FIXME", "...", ".");
+
+  private final List<String> elementParameters;
+  private final List<String> elementExceptionNames;
+  private final String mainDescription;
+  private final Map<BlockTagKey, List<String>> blockTagDescriptions;
+
+  public Javadoc(Tree tree) {
+    if (tree.is(METHOD_KINDS)) {
+      elementParameters = ((MethodTree) tree).parameters().stream()
+        .map(VariableTree::simpleName)
+        .map(IdentifierTree::name)
+        .collect(Collectors.toList());
+      elementExceptionNames = ((MethodTree) tree).throwsClauses().stream()
+        .map(Javadoc::exceptionName)
+        .filter(Objects::nonNull)
+        .collect(Collectors.toList());
+    } else if (tree.is(CLASS_KINDS)) {
+      elementParameters = ((ClassTree) tree).typeParameters().stream()
+        .map(TypeParameterTree::identifier)
+        .map(IdentifierTree::name)
+        .map(name -> "<" + name + ">")
+        .collect(Collectors.toList());
+      elementExceptionNames = Collections.emptyList();
+    } else {
+      elementParameters = Collections.emptyList();
+      elementExceptionNames = Collections.emptyList();
+    }
+
+    List<String> javadocLines = cleanLines(PublicApiChecker.getApiJavadoc(tree));
+    mainDescription = getDescription(javadocLines, -1, "");
+    blockTagDescriptions = extractBlockTags(javadocLines, Arrays.asList(BlockTag.values()));
+  }
+
+  public boolean noMainDescription() {
+    return isEmptyDescription(mainDescription);
+  }
+
+  public boolean noReturnDescription() {
+    return isEmptyDescription(blockTagDescriptions.get(BlockTagKey.of(BlockTag.RETURN, null)));
+  }
+
+  public Set<String> undocumentedParameters() {
+    return elementParameters.stream()
+      .filter(name -> isEmptyDescription(blockTagDescriptions.get(BlockTagKey.of(BlockTag.PARAM, name))))
+      .collect(Collectors.toSet());
+  }
+
+  public Set<String> undocumentedThrownExceptions() {
+    Map<String, List<String>> thrownExceptionsMap = getJavadocExceptions();
```

Comment:
Probably worth computing it only once, and store it in local private variable (to prevent being called multiple time in a row and trigger computation.)